### PR TITLE
Fix @implements emit for namespaced base types

### LINKED ISSUE #38640: Declaration file for `@implements` of type specified in `.d.ts` file includes broken syntax
**TypeScript Version:**  4.0.0-dev.20200512

**Search Terms:** implements, jsdoc, declaration

**Code**

`file.js`:
```js
/**
 * @implements {Global.Dispatcher}
 */
export class Dispatcher {
  foobar() {
    return 42;
  };
}
```

`defs.d.ts`:
```ts
declare namespace Global {
  export interface Dispatcher {
    foobar(): number;
  }
}
```

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "module": "esnext",
    "target": "esnext",
    "outDir": "lib",
    "declaration": true,
    "allowJs": true,
    "checkJs": true
  },
  "files": [
    "file.js",
    "defs.d.ts"
  ]
}
```

**Expected behavior:**
Output declaration file `file.d.ts`:
```ts
/**
 * @implements {Global.Dispatcher}
 */
export class Dispatcher implements Global.Dispatcher {
    foobar(): number;
}
export {};

```
**Actual behavior:**
```ts
declare const Dispatcher_base: undefined;
/**
 * @implements {Global.Dispatcher}
 */
export class Dispatcher implements Dispatcher_base {
    foobar(): number;
}
export {};
```

The syntax in this declaration file is broken and shows the following error:

```
'Dispatcher_base' refers to a value, but is being used as a type here.ts(2749)
```

**Playground Link:** n/a

**Related Issues:** #35629 implemented in #36292 CC @sandersn @dragomirtitian

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #38640

Couple of outstanding questions:

1. Whether an inaccessible base type is possible -- currently we only allow type references here, which is fine for chrome-devtools, but doesn't work for the pure JS case where somebody wants to declare that they implement a class that they acquired via `require` -- `require` doesn't create a real alias today, but a value, which cannot be used for `implements`.
2. `serializeImplementedType` contains a duplicate of `trySerializeTypeReference`, *except* that it calls `isTypeSymbolAccessible` instead of `getSymbolAccessibilityChain`. Both should call the same function, and then the code should be de-duped. But I'm not at all sure whether `getSymbolAccessibilityChain` should be expanded or whether both should call of `isSymbolAccessible`.

If (1) is impossible, then I can delete a lot of error reporting machinery that I currently haven't found a way to test.

Here's some more detail:

## Inaccessible base type ##

`@implements`' type reference doesn't seem to add `Value` to name lookup the way the rest of jsdoc lookups do. I'm not sure why yet. However, chrome-devtools is referring to types from a .d.ts or global classes, as far as I know. However, the following code should probably work too:

```js
const C = require('m')
/** @implements {C} */
class D { }
```

I just don't think there's much demand for that, considering that I didn't observe much usage of `@implements` outside closure codebases, and if people write *new* jsdoc, they can work around the limitation for now by creating a type alias for the type of the class: `@typedef C {typeof import('m').C}.`

## Duplication in serializeImplementedType ##

I'm going to try switching `trySerializeTypeReference` to `isValueSymbolAccessible`, and pass `allowModules: false`.

### CODE REVIEW

**Comment by sandersn** [comment]
@weswigham I decided to try isSymbolAccessible for everything, and it passes all of suite 0. What additional tests even use d.ts emit? We don't have good coverage here.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -6841,23 +6850,32 @@ namespace ts {
...
                         return createExpressionWithTypeArguments(typeArgs, reference);
                     }
                 }
 
+                function serializeImplementedType(t: Type) {
+                    const ref = trySerializeAsTypeReference(t, SymbolFlags.Type);
+                    if (ref) {
+                        return ref;
+                    }
+                    return createExpressionWithTypeArguments(/*typeArgs*/ undefined, symbolToExpression(t.symbol, context, SymbolFlags.Type));
```
Comment (weswigham) [inline src/compiler/checker.ts:6876]: If `t.symbol` is `undefined` (which is often can be), I'm pretty sure this crashes? Or makes a node which'll crash later?

**Comment by weswigham** [comment]
> What additional tests even use d.ts emit?

JS `.d.ts` emit is just the relevant conformance folder, mostly. It has a fair quantity of tests in it; but the area to cover is massive.

**Comment by TimvdLippe** [comment]
This PR appears to be approved, but not merged yet. Is there anything I could help out with to get this over the line? Not sure how best we could test this in our repository, I assume we need it packed for that?

**Comment by sandersn** [comment]
Sorry, casualty of some urgent internal work that came up. I'll see if I can address @weswigham's comment about undefined `t.symbol`.

**Pull request merged**